## Exploratory Data Analysis 

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import pickle
import functions as fn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
import numpy as np

In [2]:
car_df = pd.read_csv('data/cardataset.zip')

In [3]:
car_df.shape

(11914, 16)

In [4]:
car_df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [5]:
car_df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

subsetting the dataset to 7920 rows from 11914 total rows

In [6]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
Make                 11914 non-null object
Model                11914 non-null object
Year                 11914 non-null int64
Engine Fuel Type     11911 non-null object
Engine HP            11845 non-null float64
Engine Cylinders     11884 non-null float64
Transmission Type    11914 non-null object
Driven_Wheels        11914 non-null object
Number of Doors      11908 non-null float64
Market Category      8172 non-null object
Vehicle Size         11914 non-null object
Vehicle Style        11914 non-null object
highway MPG          11914 non-null int64
city mpg             11914 non-null int64
Popularity           11914 non-null int64
MSRP                 11914 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 1.5+ MB


In [7]:
# plt.hist(car_df['Year']);
car_df1 = car_df.loc[car_df['Year'] > 2009]

In [8]:
car_df1.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        0
Engine HP              63
Engine Cylinders       20
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      1642
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [9]:
car_df1.dropna(subset=['Number of Doors'], axis=0, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [10]:
car_df1.dropna(subset=['Engine HP'], axis=0, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [11]:
car_df1.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        0
Engine HP               0
Engine Cylinders       19
Transmission Type       0
Driven_Wheels           0
Number of Doors         0
Market Category      1637
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [12]:
car_df1.loc[car_df1['Model']== 'RX-8']

Make Model  Year             Engine Fuel Type  Engine HP  \
8705  Mazda  RX-8  2010  premium unleaded (required)      232.0   
8706  Mazda  RX-8  2010  premium unleaded (required)      212.0   
8707  Mazda  RX-8  2010  premium unleaded (required)      212.0   
8708  Mazda  RX-8  2010  premium unleaded (required)      232.0   
8709  Mazda  RX-8  2010  premium unleaded (required)      232.0   
8710  Mazda  RX-8  2011  premium unleaded (required)      212.0   
8711  Mazda  RX-8  2011  premium unleaded (required)      232.0   
8712  Mazda  RX-8  2011  premium unleaded (required)      232.0   
8713  Mazda  RX-8  2011  premium unleaded (required)      232.0   
8714  Mazda  RX-8  2011  premium unleaded (required)      212.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
8705               NaN            MANUAL  rear wheel drive              4.0   
8706               NaN         AUTOMATIC  rear wheel drive              4.0   
8707               NaN         AUTOMATIC  rear wheel drive              4.0   
8708               NaN            MANUAL  rear wheel drive              4.0   
8709               NaN            MANUAL  rear wheel drive              4.0   
8710               NaN         AUTOMATIC  rear wheel drive              4.0   
8711               NaN            MANUAL  rear wheel drive              4.0   
8712               NaN            MANUAL  rear wheel drive              4.0   
8713               NaN            MANUAL  rear wheel drive              4.0   
8714               NaN         AUTOMATIC  rear wheel drive              4.0   

     Market Category Vehicle Size Vehicle Style  highway MPG  city mpg  \
8705     Performance      Compact         Coupe           22        16   
8706     Performance      Compact         Coupe           23        16   
8707     Performance      Compact         Coupe           23        16   
8708     Performance      Compact         Coupe           22        16   
8709     Performance      Compact         Coupe           22        16   
8710     Performance      Compact         Coupe           23        16   
8711     Performance      Compact         Coupe           22        16   
8712     Performance      Compact         Coupe           22        16   
8713     Performance      Compact         Coupe           22        16   
8714     Performance      Compact         Coupe           23        16   

      Popularity   MSRP  
8705         586  32140  
8706         586  26645  
8707         586  32810  
8708         586  26645  
8709         586  32110  
8710         586  32960  
8711         586  32260  
8712         586  32290  
8713         586  26795  
8714         586  26795

In [13]:
car_df1.loc[car_df1['Engine Cylinders'].isna()] = car_df1.loc[car_df1['Engine Cylinders'].isna()].fillna(0)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [14]:
car_df1.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        0
Engine HP               0
Engine Cylinders        0
Transmission Type       0
Driven_Wheels           0
Number of Doors         0
Market Category      1637
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [15]:
car_df1.loc[car_df1['Engine HP'].isna()]['Model'].value_counts()

Series([], Name: Model, dtype: int64)

In [16]:
car_df1.iloc[6578]['Engine HP'] = 200.0

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [17]:
car_df1.loc[car_df1['Model']=='Model S']

Empty DataFrame
Columns: [Make, Model, Year, Engine Fuel Type, Engine HP, Engine Cylinders, Transmission Type, Driven_Wheels, Number of Doors, Market Category, Vehicle Size, Vehicle Style, highway MPG, city mpg, Popularity, MSRP]
Index: []

In [18]:
car_df1.loc[car_df1['Market Category'].isna()]['Make'].value_counts()

Toyota        328
Nissan        197
Chevrolet     174
Volkswagen    170
Honda         139
Ford          124
GMC            76
Suzuki         68
Hyundai        62
Kia            59
Dodge          56
Mazda          44
Mitsubishi     30
Subaru         30
Chrysler       29
FIAT           21
Buick          17
Scion          12
Pontiac         1
Name: Make, dtype: int64

In [19]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7856 entries, 0 to 11912
Data columns (total 16 columns):
Make                 7856 non-null object
Model                7856 non-null object
Year                 7856 non-null int64
Engine Fuel Type     7856 non-null object
Engine HP            7856 non-null float64
Engine Cylinders     7856 non-null float64
Transmission Type    7856 non-null object
Driven_Wheels        7856 non-null object
Number of Doors      7856 non-null float64
Market Category      6219 non-null object
Vehicle Size         7856 non-null object
Vehicle Style        7856 non-null object
highway MPG          7856 non-null int64
city mpg             7856 non-null int64
Popularity           7856 non-null int64
MSRP                 7856 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 1.0+ MB


In [20]:
car_df1.dropna(axis=0, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [21]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6219 entries, 0 to 11912
Data columns (total 16 columns):
Make                 6219 non-null object
Model                6219 non-null object
Year                 6219 non-null int64
Engine Fuel Type     6219 non-null object
Engine HP            6219 non-null float64
Engine Cylinders     6219 non-null float64
Transmission Type    6219 non-null object
Driven_Wheels        6219 non-null object
Number of Doors      6219 non-null float64
Market Category      6219 non-null object
Vehicle Size         6219 non-null object
Vehicle Style        6219 non-null object
highway MPG          6219 non-null int64
city mpg             6219 non-null int64
Popularity           6219 non-null int64
MSRP                 6219 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 826.0+ KB


In [22]:
car_df1.duplicated().sum()

241

In [23]:
car_df1.drop_duplicates(inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [24]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5978 entries, 0 to 11912
Data columns (total 16 columns):
Make                 5978 non-null object
Model                5978 non-null object
Year                 5978 non-null int64
Engine Fuel Type     5978 non-null object
Engine HP            5978 non-null float64
Engine Cylinders     5978 non-null float64
Transmission Type    5978 non-null object
Driven_Wheels        5978 non-null object
Number of Doors      5978 non-null float64
Market Category      5978 non-null object
Vehicle Size         5978 non-null object
Vehicle Style        5978 non-null object
highway MPG          5978 non-null int64
city mpg             5978 non-null int64
Popularity           5978 non-null int64
MSRP                 5978 non-null int64
dtypes: float64(3), int64(5), object(8)
memory usage: 794.0+ KB


In [25]:
# car_df1.loc[(car_df1['Make']== 'FIAT') & (car_df1['Model']== '500')]

In [26]:
car_df1.Make.unique()

array(['BMW', 'FIAT', 'Chrysler', 'Mazda', 'Nissan', 'Ferrari',
       'Alfa Romeo', 'McLaren', 'Maybach', 'Porsche', 'Toyota', 'Saab',
       'Audi', 'GMC', 'Hyundai', 'Honda', 'Mercedes-Benz', 'Cadillac',
       'Chevrolet', 'Dodge', 'Lamborghini', 'Bentley', 'Volkswagen',
       'Subaru', 'Ford', 'Volvo', 'Kia', 'Lincoln', 'Lexus',
       'Rolls-Royce', 'Aston Martin', 'Land Rover', 'Mitsubishi', 'Lotus',
       'Buick', 'Infiniti', 'Scion', 'Genesis', 'Maserati', 'Suzuki',
       'HUMMER', 'Acura', 'Pontiac'], dtype=object)

In [27]:
# with open('car_df1.pickle', 'wb') as f:
#     pickle.dump(car_df1, f, pickle.HIGHEST_PROTOCOL)

create a new column to assign carmaker origin to each row 

In [28]:
with open('make_origin.pickle', 'rb') as f:
    make_origin_dict = pickle.load(f)

In [29]:
make_origin_dict

{'BMW': 'Germany',
 'FIAT': 'Italy',
 'Chrysler': 'USA',
 'Mazda': 'Japan',
 'Nissan': 'Japan',
 'Ferrari': 'Italy',
 'Alfa Romeo': 'Italy',
 'McLaren': 'UK',
 'Maybach': 'Germany',
 'Porsche': 'Germany',
 'Toyota': 'Japan',
 'Saab': 'Sweden',
 'Audi': 'Germany',
 'GMC': 'USA',
 'Hyundai': 'Korea',
 'Honda': 'Japan',
 'Mercedes-Benz': 'Germany',
 'Cadillac': 'USA',
 'Chevrolet': 'USA',
 'Dodge': 'USA',
 'Lamborghini': 'Italy',
 'Bentley': 'UK',
 'Volkswagen': 'Germany',
 'Subaru': 'Japan',
 'Ford': 'USA',
 'Volvo': 'Sweden',
 'Kia': 'Korea',
 'Lincoln': 'USA',
 'Lexus': 'Japan',
 'Rolls-Royce': 'UK',
 'Aston Martin': 'UK',
 'Land Rover': 'UK',
 'Mitsubishi': 'Japan',
 'Lotus': 'UK',
 'Buick': 'USA',
 'Infiniti': 'Japan',
 'Scion': 'Japan',
 'Genesis': 'Korea',
 'Maserati': 'Italy',
 'Suzuki': 'Japan',
 'HUMMER': 'USA',
 'Acura': 'Japan',
 'Pontiac': 'USA'}

split the attributes in the column market catergory to seperate columns

In [30]:
car_df1 = fn.market_columns(car_df1)

/Users/Erica/flatiron/mod_5/mod_5_project/car-classification-using-machine-learning/functions.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  df[col] = df['Market Category'].apply(lambda x: 1 if col in x.split(',') else 0)
/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [31]:
car_df1.drop(columns='Number of Doors', axis=1, inplace=True)

/Users/Erica/anaconda3/envs/learn-env/lib/python3.6/site-packages/pandas/core/frame.py:3697: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [32]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5978 entries, 0 to 11912
Data columns (total 24 columns):
Make                 5978 non-null object
Model                5978 non-null object
Year                 5978 non-null int64
Engine Fuel Type     5978 non-null object
Engine HP            5978 non-null float64
Engine Cylinders     5978 non-null float64
Transmission Type    5978 non-null object
Driven_Wheels        5978 non-null object
Vehicle Size         5978 non-null object
Vehicle Style        5978 non-null object
highway MPG          5978 non-null int64
city mpg             5978 non-null int64
Popularity           5978 non-null int64
MSRP                 5978 non-null int64
Exotic               5978 non-null int64
Flex Fuel            5978 non-null int64
Performance          5978 non-null int64
Diesel               5978 non-null int64
Crossover            5978 non-null int64
Hybrid               5978 non-null int64
Hatchback            5978 non-null int64
Luxury              

In [33]:
car_df1['Vehicle Style'].value_counts()

4dr SUV                1816
Sedan                  1401
Coupe                   667
4dr Hatchback           557
Convertible             455
Crew Cab Pickup         260
Wagon                   225
2dr Hatchback           189
Extended Cab Pickup     119
Regular Cab Pickup      105
Passenger Van           100
Passenger Minivan        38
Cargo Van                34
2dr SUV                   6
Convertible SUV           5
Cargo Minivan             1
Name: Vehicle Style, dtype: int64

In [34]:
car_df1 = car_df1.loc[(car_df1['Vehicle Style'] != '2dr SUV') 
                      & (car_df1['Vehicle Style'] != 'Convertible SUV') &
         (car_df1['Vehicle Style'] != 'Cargo Minivan')]

In [35]:
car_df1['Driven_Wheels'].value_counts()

front wheel drive    1944
all wheel drive      1841
rear wheel drive     1679
four wheel drive      502
Name: Driven_Wheels, dtype: int64

In [36]:
car_df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5966 entries, 0 to 11912
Data columns (total 24 columns):
Make                 5966 non-null object
Model                5966 non-null object
Year                 5966 non-null int64
Engine Fuel Type     5966 non-null object
Engine HP            5966 non-null float64
Engine Cylinders     5966 non-null float64
Transmission Type    5966 non-null object
Driven_Wheels        5966 non-null object
Vehicle Size         5966 non-null object
Vehicle Style        5966 non-null object
highway MPG          5966 non-null int64
city mpg             5966 non-null int64
Popularity           5966 non-null int64
MSRP                 5966 non-null int64
Exotic               5966 non-null int64
Flex Fuel            5966 non-null int64
Performance          5966 non-null int64
Diesel               5966 non-null int64
Crossover            5966 non-null int64
Hybrid               5966 non-null int64
Hatchback            5966 non-null int64
Luxury              

In [37]:
car_df1['Origin'] = car_df1['Make'].apply(lambda m: make_origin_dict[m])

In [38]:
car_df1.drop(['Make', 'Model'], axis=1, inplace=True)

In [39]:
X = car_df1.drop('Origin', axis=1)
y = car_df1['Origin']

In [40]:
y

0        Germany
1        Germany
2        Germany
3        Germany
4        Germany
5        Germany
6        Germany
7        Germany
8        Germany
9        Germany
10       Germany
11       Germany
12       Germany
13       Germany
15       Germany
16       Germany
32         Italy
33         Italy
34         Italy
41       Germany
42       Germany
43       Germany
44       Germany
45       Germany
46       Germany
47       Germany
48       Germany
49       Germany
50       Germany
51       Germany
          ...   
11865        USA
11866        USA
11869        USA
11870        USA
11871        USA
11872        USA
11873        USA
11874        USA
11875        USA
11876        USA
11877        USA
11878        USA
11879        USA
11880        USA
11894    Germany
11895    Germany
11896    Germany
11897    Germany
11898    Germany
11899    Germany
11900    Germany
11901    Germany
11902    Germany
11906      Japan
11907      Japan
11908      Japan
11909      Japan
11910      Jap

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [42]:
# origins = list(set(make_origin_dict.values()))

# origin_code = {}
# for i in range(len(origins)):
#     origin_code[origins[i]] = i
# origin_code

# car_df1['origin_code'] = car_df1['Origin'].apply(lambda x: origin_code[x])
# car_df1.drop(columns='Origin', axis=1, inplace=True)

In [43]:
origins = list(set(make_origin_dict.values()))

origin_code = {}
for i in range(len(origins)):
    origin_code[origins[i]] = i
#origin_code

y_train = y_train.apply(lambda x: origin_code[x])

In [44]:
y_test = y_test.apply(lambda x: origin_code[x])

In [45]:
X

Year                Engine Fuel Type  Engine HP  Engine Cylinders  \
0      2011     premium unleaded (required)      335.0               6.0   
1      2011     premium unleaded (required)      300.0               6.0   
2      2011     premium unleaded (required)      300.0               6.0   
3      2011     premium unleaded (required)      230.0               6.0   
4      2011     premium unleaded (required)      230.0               6.0   
5      2012     premium unleaded (required)      230.0               6.0   
6      2012     premium unleaded (required)      300.0               6.0   
7      2012     premium unleaded (required)      300.0               6.0   
8      2012     premium unleaded (required)      230.0               6.0   
9      2013     premium unleaded (required)      230.0               6.0   
10     2013     premium unleaded (required)      300.0               6.0   
11     2013     premium unleaded (required)      230.0               6.0   
12     2013     premium unleaded (required)      300.0               6.0   
13     2013     premium unleaded (required)      230.0               6.0   
15     2013     premium unleaded (required)      320.0               6.0   
16     2013     premium unleaded (required)      320.0               6.0   
32     2017  premium unleaded (recommended)      160.0               4.0   
33     2017  premium unleaded (recommended)      160.0               4.0   
34     2017  premium unleaded (recommended)      160.0               4.0   
41     2016     premium unleaded (required)      240.0               4.0   
42     2016     premium unleaded (required)      240.0               4.0   
43     2016     premium unleaded (required)      320.0               6.0   
44     2016     premium unleaded (required)      240.0               4.0   
45     2016     premium unleaded (required)      240.0               4.0   
46     2016     premium unleaded (required)      320.0               6.0   
47     2016     premium unleaded (required)      240.0               4.0   
48     2016     premium unleaded (required)      320.0               6.0   
49     2016     premium unleaded (required)      320.0               6.0   
50     2017  premium unleaded (recommended)      335.0               6.0   
51     2017  premium unleaded (recommended)      335.0               6.0   
...     ...                             ...        ...               ...   
11865  2015        flex-fuel (unleaded/E85)      355.0               8.0   
11866  2015        flex-fuel (unleaded/E85)      355.0               8.0   
11869  2016  premium unleaded (recommended)      420.0               8.0   
11870  2016        flex-fuel (unleaded/E85)      355.0               8.0   
11871  2016  premium unleaded (recommended)      420.0               8.0   
11872  2016        flex-fuel (unleaded/E85)      355.0               8.0   
11873  2016        flex-fuel (unleaded/E85)      355.0               8.0   
11874  2016        flex-fuel (unleaded/E85)      355.0               8.0   
11875  2017  premium unleaded (recommended)      420.0               8.0   
11876  2017        flex-fuel (unleaded/E85)      355.0               8.0   
11877  2017        flex-fuel (unleaded/E85)      355.0               8.0   
11878  2017  premium unleaded (recommended)      420.0               8.0   
11879  2017        flex-fuel (unleaded/E85)      355.0               8.0   
11880  2017        flex-fuel (unleaded/E85)      355.0               8.0   
11894  2014     premium unleaded (required)      240.0               4.0   
11895  2014     premium unleaded (required)      300.0               6.0   
11896  2014     premium unleaded (required)      335.0               6.0   
11897  2015     premium unleaded (required)      240.0               4.0   
11898  2015     premium unleaded (required)      300.0               6.0   
11899  2015     premium unleaded (required)      335.0               6.0   
11900  2016     premium unleaded (required)      300.0 

In [46]:
X.columns

Index(['Year', 'Engine Fuel Type', 'Engine HP', 'Engine Cylinders',
       'Transmission Type', 'Driven_Wheels', 'Vehicle Size', 'Vehicle Style',
       'highway MPG', 'city mpg', 'Popularity', 'MSRP', 'Exotic', 'Flex Fuel',
       'Performance', 'Diesel', 'Crossover', 'Hybrid', 'Hatchback', 'Luxury',
       'Factory Tuner', 'High-Performance'],
      dtype='object')

In [47]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5966 entries, 0 to 11912
Data columns (total 22 columns):
Year                 5966 non-null int64
Engine Fuel Type     5966 non-null object
Engine HP            5966 non-null float64
Engine Cylinders     5966 non-null float64
Transmission Type    5966 non-null object
Driven_Wheels        5966 non-null object
Vehicle Size         5966 non-null object
Vehicle Style        5966 non-null object
highway MPG          5966 non-null int64
city mpg             5966 non-null int64
Popularity           5966 non-null int64
MSRP                 5966 non-null int64
Exotic               5966 non-null int64
Flex Fuel            5966 non-null int64
Performance          5966 non-null int64
Diesel               5966 non-null int64
Crossover            5966 non-null int64
Hybrid               5966 non-null int64
Hatchback            5966 non-null int64
Luxury               5966 non-null int64
Factory Tuner        5966 non-null int64
High-Performance     5

In [48]:
# Create train and test dataframes with only the categorical variables

X_train_obj = X_train[[col for col,dtype in list(zip(X_train.columns, X_train.dtypes)) 
                          if dtype == np.dtype('O')]]
X_test_obj = X_test[[col for col,dtype in list(zip(X_test.columns, X_test.dtypes)) 
                        if dtype == np.dtype('O')]]

In [49]:
X_train_nonobj = X_train[[col for col,dtype in list(zip(X_train.columns, X_train.dtypes)) 
                          if dtype != np.dtype('O')]]
X_test_nonobj = X_test[[col for col,dtype in list(zip(X_test.columns, X_test.dtypes)) 
                        if dtype != np.dtype('O')]]

In [50]:
import numpy as np
ohe = OneHotEncoder(handle_unknown='ignore')
X_train_obj_ohe = ohe.fit_transform(X_train_obj)
X_test_obj_ohe = ohe.transform(X_test_obj)

In [51]:
X_train_nonobj_df = pd.DataFrame(X_train_nonobj).reset_index(drop=True) 
X_train_obj_ohe_df = pd.DataFrame(X_train_obj_ohe.todense(), columns=ohe.get_feature_names()).reset_index(drop=True)

In [52]:
X_train_all = pd.concat([X_train_nonobj_df, X_train_obj_ohe_df], axis=1)

In [53]:
X_test_nonobj_df = pd.DataFrame(X_test_nonobj).reset_index(drop=True) 
X_test_obj_ohe_df = pd.DataFrame(X_test_obj_ohe.todense(), columns=ohe.get_feature_names()).reset_index(drop=True)

In [54]:
X_test_all = pd.concat([X_train_nonobj_df, X_train_obj_ohe_df], axis=1)

In [55]:
X_train_all = pd.concat([pd.DataFrame(X_train_nonobj), pd.DataFrame(X_train_obj_ohe.todense(), columns=ohe.get_feature_names())], axis=1)
X_test_all = pd.concat([pd.DataFrame(X_test_nonobj), pd.DataFrame(X_test_obj_ohe.todense(), columns=ohe.get_feature_names())], axis=1)

In [56]:
# cat_features = ['Engine Fuel Type', 'Transmission Type', 'Driven_Wheels', 'Vehicle Size', 'Vehicle Style']
# cat_index=[]
# for ind, val in enumerate(X_train.columns):
#     #print(ind, val)
#     if val in cat_features:
#         cat_index.append(ind)

In [57]:
# X_train_df = pd.DataFrame(X_train_ohe.todense(), columns=ohe.get_feature_names())
# X_train_df.columns